### PR TITLE
feat: Download tickets as pdf

### LINKED ISSUE #1338: Download Tickets as PDF
**Describe the feature you'd like**
There should be an option to download ticket in pdf format 

**Would you like to work on the issue?**
- [x] Yes

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
Fixes #1338 

Changes: Added Download Button to download tickets both landscape and portrait

Screenshots for the change:screenshot-1553118209746

### CODE REVIEW

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal should i add a broadcast receiver to notify user once the ticket has been downloaded and also should i add check if the file is already downloaded don't download it again?

**Comment by iamareebjamal** [comment]
Simply open the link in the browser

**Comment by aggarwalpulkit596** [comment]
But for that i need to add a broadcast reciever or you're suggesting to just open the pdf url in brower instead of downloading

**Comment by iamareebjamal** [comment]
2nd one

**Comment by aggarwalpulkit596** [comment]
But as a user perspective wouldn't it be  odd to download file we are opening a web browser and we can do that without opening the brower

**Comment by iamareebjamal** [comment]
I may want to just see the PDF, or download it with a special Downloader than the default download manager. I hate when app developers force me to choose their way of doing things when Android has had a robust intent system from version 1

**Comment by aggarwalpulkit596** [comment]
Cool I'll just pass the url in brower intent

**Comment by aggarwalpulkit596** [comment]
Cool I'll just pass the url in brower intent

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal can i add anko commons for intents ?

**Comment by iamareebjamal** [comment]
Yes you can

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt [hunk 1]
```diff
@@ -5,18 +5,7 @@ import android.net.Uri
...
-import kotlinx.android.synthetic.main.item_card_order_details.view.eventName
-import kotlinx.android.synthetic.main.item_card_order_details.view.eventSummary
-import kotlinx.android.synthetic.main.item_card_order_details.view.location
-import kotlinx.android.synthetic.main.item_card_order_details.view.map
-import kotlinx.android.synthetic.main.item_card_order_details.view.name
-import kotlinx.android.synthetic.main.item_card_order_details.view.orderIdentifier
-import kotlinx.android.synthetic.main.item_card_order_details.view.organizer
-import kotlinx.android.synthetic.main.item_card_order_details.view.qrCodeView
-import kotlinx.android.synthetic.main.item_card_order_details.view.organizerLabel
+import kotlinx.android.synthetic.main.item_card_order_details.view.*
```
Comment (theLimitBreaker) [inline app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: don't use `WildCard` imports . That fails your spotlesscheck

**Comment by aggarwalpulkit596** [comment]
That isn't a wildcard import all of the views was including in the import so it was converted into * and travis build is also without any erros otherwise there would have been a error

**Comment by theLimitBreaker** [comment]
You could see this for wildcard imports : https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

**Comment by liveHarshit** [comment]
@iamareebjamal Ticket pdf should mail to the user. Is this feature not implemented on the server yet?

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt [hunk 2]
```diff
@@ -56,6 +46,12 @@ class OrderDetailsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
                     itemView.context.startActivity(mapIntent)
                 }
             }
+            if (attendee.pdfUrl!!.isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: Please do not use non null assertions.

[the remark this answers is not in the document]
Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: okayy 
will remove it

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated please review

File: app/build.gradle [hunk 3]
```diff
@@ -93,6 +93,8 @@ dependencies {
     def roomVersion = "2.1.0-alpha04"
     def ktx_version = "1.0.0"
     def ktx2_version = "2.0.0"
+    def anko_version = "0.10.6"
+
```
Comment (nikit19) [inline app/build.gradle]: extra

File: app/build.gradle [hunk 4]
```diff
@@ -161,6 +163,10 @@ dependencies {
     implementation 'android.arch.navigation:navigation-fragment-ktx:1.0.0'
     implementation 'android.arch.navigation:navigation-ui-ktx:1.0.0'
 
+    //Anko
+    implementation "org.jetbrains.anko:anko-commons:$anko_version"
+
+
```
Comment (nikit19) [inline app/build.gradle]: extra

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -7,6 +7,7 @@
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
 
+
```
Comment (nikit19) [inline app/src/main/AndroidManifest.xml]: extra

**Comment by iamareebjamal** [comment]
@liveHarshit It is implemented AFAIK

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsRecyclerAdapter.kt [hunk 6]
```diff
@@ -25,8 +26,9 @@ class OrderDetailsRecyclerAdapter : RecyclerView.Adapter<OrderDetailsViewHolder>
         this.event = event
     }
 
-    fun setListener(listener: EventDetailsListener) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsRecyclerAdapter.kt]: Create another function for the listener.

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

**Comment by aggarwalpulkit596** [comment]
Where should i use lamba?

**Comment by iamareebjamal** [comment]
While setting the listener

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

**Comment by aggarwalpulkit596** [comment]
@nikit19 anything else ?

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt (hunk 1 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: Revert the wildcard import

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsFragment.kt [hunk 7]
```diff
@@ -85,7 +86,14 @@ class OrderDetailsFragment : Fragment() {
             }
         }
 
+        val downloadClickListener = object : OrderDetailsRecyclerAdapter.DownloadClickListener {
+            override fun onClick(url: String) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsFragment.kt]: Use lambda

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

**Comment by iamareebjamal** [comment]
Still not using lambda

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal i'm not sure how to use lamba where you're asking me to use that i've used lamba while setting onclick listerner but while instating an object i'm not sure

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal 
https://github.com/fossasia/open-event-android/blob/88dd47cf42d2e446d97e786aad7e10b80afa0dd6/app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsFragment.kt#L77
even here is it using same pattern i'm not sure that i'll be able to lamba expression here

**Comment by iamareebjamal** [comment]
```kotlin
adapter.setOnBlahListener { url ->
  context?.browse(url)
}
```

This is called a lamda

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal i know that but i can't do that while initializing it to variable

**Comment by iamareebjamal** [comment]
No one asked to store the listener in a variable

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal you're suggesting that i should move clicklisterner to view holder class ?

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt [hunk 8]
```diff
@@ -56,6 +58,12 @@ class OrderDetailsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
                     itemView.context.startActivity(mapIntent)
                 }
             }
+            if (!attendee.pdfUrl.isNullOrBlank()) {
+                itemView.downloadButton.isEnabled = true
+                itemView.downloadButton.setOnClickListener { it ->
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: No use of it -> if you are not going to rename it.

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt [hunk 9]
```diff
@@ -56,6 +58,12 @@ class OrderDetailsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
...
+                    it.context.browse(attendee.pdfUrl)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: Use itemView.context or context

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -108,7 +109,7 @@
     <string name="subtotal">Subtotal</string>
     <string name="view">(view)</string>
     <string name="hide">(hide)</string>
-    <string name="no_tickets_message">Invalid Quantity. Please enter a quantity of 1 or more</string>
+    <string name="no_tickets_message">Invalid Quantity. Please enter a quantity of 1 and more</string>
```
Comment (iamareebjamal) [inline app/src/main/res/values/strings.xml]: Revert

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

File: app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt [hunk 11]
```diff
@@ -56,6 +58,12 @@ class OrderDetailsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
                     itemView.context.startActivity(mapIntent)
                 }
             }
+            if (!attendee.pdfUrl.isNullOrBlank()) {
+                itemView.downloadButton.isEnabled = true
+                itemView.downloadButton.setOnClickListener { downloadBtn ->
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/order/OrderDetailsViewHolder.kt]: downloadBtn -> is unneeded. Remove

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal completed

**Comment by ShridharGoel** [comment]
@aggarwalpulkit596 Please fix PR title so that Semantic passes.

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal anything left out ?

**Comment by aman210697** [comment]
> @iamareebjamal anything left out ?

Add space after colon in PR title, to pass the semantic check

**Pull request merged**